pkg/metrics-cache/metricscache.go

```diff
@@ -0,0 +1,153 @@
+package metricscache
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/armon/go-metrics"
+)
+
+type metric struct {
+	key    []string
+	val    float32
+	labels []metrics.Label
+}
+
+// Sink is a temporary sink that caches metrics until a real sink is set in SetSink.
+// it implements the metrics.MetricSink interface
+type Sink struct {
+	gauges   []metric
+	counters []metric
+	samples  []metric
+	keys     []metric
+
+	realSink metrics.MetricSink
+
+	mu        sync.Mutex
+	checkLock *atomic.Bool
+	once      sync.Once
+}
+
+// NewSink returns a pointer to a sink with empty cache
+func NewSink() *Sink {
+	checkLock := &atomic.Bool{}
+	checkLock.Store(true)
+
+	return &Sink{
+		gauges:    []metric{},
+		counters:  []metric{},
+		samples:   []metric{},
+		keys:      []metric{},
+		mu:        sync.Mutex{}, // this lock is used to control access around the slices of metrics above
+		checkLock: checkLock,    // we only need to check the lock if we haven't yet set the real sink
+	}
+}
+
+// SetGauge defaults to SetGaugeWithLabels
+func (s *Sink) SetGauge(key []string, val float32) {
+	s.SetGaugeWithLabels(key, val, nil)
+}
+
+// SetGaugeWithLabels sends metrics to the real sink otherwise caches them
+func (s *Sink) SetGaugeWithLabels(key []string, val float32, labels []metrics.Label) {
+	if ok := s.checkLock.Load(); ok {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+	}
+
+	if s.realSink != nil {
+		s.realSink.SetGaugeWithLabels(key, val, labels)
+		return
+	}
+
+	s.gauges = append(s.gauges, metric{key: key, val: val, labels: labels})
```

Comment:
non-blocking: What do you think about changing `realSink` to an `atomic.Value` and removing `checkLock`? Then I think these methods could look something like this.

```suggestion
	if sink := s.realSink.Load().(metrics.MetricsSink); sink != nil {
		sink.SetGaugeWithLabels(key, val, labels)
		return
	}
	
	s.mu.Lock()
	defer s.mu.Unlock()
	s.gauges = append(s.gauges, metric{key: key, val: val, labels: labels})
```

Reply:
Yeah Matt and I had that as the original design but the problem with that there can be a race when we `SetSink` and get the lock and when the realSink is initialized. A metric could be waiting for the lock and then unblock and append to one of the slices and then we lose that metric. It's not the end of the world so I am willing to sacrifice a few metrics for readability but it is possible it could result in a bit of a dip in metrics everything we call `SetSink`. 🤷  open to your thoughts.

Reply:
Ah, okay. That makes sense.